dev/AppNotifications/AppNotificationManager.cpp

```diff
@@ -0,0 +1,358 @@
+        if (!PushNotificationHelpers::IsPackagedAppScenario())
+        {
+            std::wstring notificationId{ RetrieveNotificationAppId() };
+            if (!AppModel::Identity::IsPackagedProcess())
+            {
+                THROW_IF_FAILED(PushNotifications_RegisterFullTrustApplication(notificationId.c_str(), GUID_NULL));
+
+                storedComActivatorString = RegisterComActivatorGuidAndAssets();
+                // Remove braces around the guid string
+                storedComActivatorString = storedComActivatorString.substr(1, storedComActivatorString.size() - 2);
+            }
+
+            wil::unique_cotaskmem_string processName;
+            THROW_IF_FAILED(GetCurrentProcessPath(processName));
+            auto notificationPlatform{ PushNotificationHelpers::GetNotificationPlatform() };
+            THROW_IF_FAILED(notificationPlatform->AddToastRegistrationMapping(processName.get(), notificationId.c_str()));
+        }
+
+        winrt::guid registeredClsid{ GUID_NULL };
+        if (AppModel::Identity::IsPackagedProcess())
+        {
+            THROW_IF_FAILED(PushNotificationHelpers::GetComRegistrationFromRegistry(expectedAppServerArgs.data(), registeredClsid));
+        }
+
+        // Create event handle before COM Registration otherwise if a notification arrives will lead to race condition
+        GetWaitHandleForArgs().create();
+
+        {
+            auto lock{ m_lock.lock_exclusive() };
+            THROW_HR_IF_MSG(E_INVALIDARG, m_notificationComActivatorRegistration, "Already Registered for App Notifications!");
+
+            // Check if the caller has registered event handlers, if so the REGCLS_MULTIPLEUSE flag will cause COM to ensure that all activators
+            // are routed inproc, otherwise with REGCLS_MULTIPLEUSE COM will launch a new process of the Win32 app for each invocation.
+            THROW_IF_FAILED(::CoRegisterClassObject(
+                AppModel::Identity::IsPackagedProcess() ? registeredClsid : winrt::guid(storedComActivatorString),
+                winrt::make<AppNotificationManagerFactory>().get(),
+                CLSCTX_LOCAL_SERVER,
+                m_notificationHandlers ? REGCLS_MULTIPLEUSE : REGCLS_SINGLEUSE,
+                &m_notificationComActivatorRegistration));
+        }
+    }
+
+    void AppNotificationManager::Unregister()
+    {
+        auto lock{ m_lock.lock_exclusive() };
+        THROW_HR_IF_MSG(HRESULT_FROM_WIN32(ERROR_NOT_FOUND), !m_notificationComActivatorRegistration, "Not Registered for App Notifications!");
+        m_notificationComActivatorRegistration.reset();
+    }
+
+    void AppNotificationManager::UnregisterAll()
+    {
+        Unregister();
+
+        // Remove any Registrations from the Long Running Process that are necessary for Cloud toasts
+        if (!PushNotificationHelpers::IsPackagedAppScenario())
+        {
+            wil::unique_cotaskmem_string processName;
+            THROW_IF_FAILED(GetCurrentProcessPath(processName));
```

Comment:
Nit: maybe add a future WIL helper for `wil::GetCurrentProcessPath()` that returns this value?

Reply:
Who should we @ for this?

Reply:
https://github.com/microsoft/WindowsAppSDK/issues/2223